Load() loads incorrect extension when specified

GuiMacielPereira opened the pull request:
### Description of work
As described in #35661 the Load() algorithm might load an incorrect file extension even when the extension is specified in the file name.

#### Summary of work
I have maintained the same behaviour for backwards compatibility, the only change I made is that if the file with the specified extension is not found, a warning is issued to let the user know that other extensions will be considered.

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

### To test:
Run the code in the original issue and check that a warning is displayed in the console log. Load the same file without an extension specified and check that no warning is displayed.

Turn on the debug level on the console log and run the code in the original issue again. Check that at first only the extension specified in the filename is considered and then all extensions are considered.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

GuiMacielPereira committed:
Look for specified extension only

When the filename contains an extension, the FileFinder will
now fail to find the file.

There are four options:

- filename contains extension + useOnlyExts=True: searches exts + extension (filename)
- filename contains extension + useOnlyExts=False: searches only extension (filename)
- filename does not contain extension + useOnlyExts=True: searches only exts
- filename does not contain extension + useOnlyExts=False: searches exts + extensions (facility)

GuiMacielPereira committed:
Add comments

GuiMacielPereira committed:
Run search with other extensions if not found

In the case where search is run with a specified extension in
the filename, search for that extension only first.
If not found, issue a warning and try again with the
other facility extensions + extensions that were passed as
an argument.
uniqueExtensions.pop_back() avoids running search for
specified extension again.

GuiMacielPereira committed:
Fix redundant if statement and add comments

Corrected if statement after warning from cppcheck

GuiMacielPereira committed:
Fix logic to pass unit test

Unit test had failed because I had overlooked the fact that
when useExtsOnly=True, the search should fail if the specified
extensions are incorrect

GuiMacielPereira committed:
Make useExtsOnly ignore ext in filename

This behaviour was not specified in the existing unit tests,
but from what I gathered when useExtsOnly=True then only those
extensions should be considered, even if an extension is explicitly
written in the filename.

GuiMacielPereira committed:
Add unit tests

Added a few unit tests to cover behaviour that was previously ambiguous.
If useExtsOnly=True, then only those extensions should be considered.
If useExtsOnly=False, it should find the file whether the extension is
correctly specified in the filename or not.

GuiMacielPereira committed:
Update test properties with test search paths

Running the FileFinderTest.h locally doesn't use the
mantid properties meant for testing automatically, so
the search paths meant for testing are not searched on,
which causes some tests to fail.

These two lines ensure that the mantid properties for
tests are used.

GuiMacielPereira committed:
Add release note

jhaigh0 commented on Framework/API/test/FileFinderTest.h:
```diff
@@ -347,6 +349,21 @@ class FileFinderTest : public CxxTest::TestSuite {
   }
 
+  void testFindRunWithOverwriteExtensionsAndCorrectExtensionInFilename() {
+    std::string path;
+
+    // This file is .nxs or .RAW
+    const std::vector<std::string> incorrect_extension = {".txt"};
+    path = FileFinder::Instance().findRun("MUSR15189.nxs", incorrect_extension, true).result();
+    TS_ASSERT_EQUALS(path, "");
```
```suggestion
    // This file is .nxs or .RAW
    const std::vector<std::string> incorrect_extension = {".txt"};
    std::string path = FileFinder::Instance().findRun("MUSR15189.nxs", incorrect_extension, true).result();
    TS_ASSERT_EQUALS(path, "");
```

jhaigh0 commented on Framework/API/src/FileFinder.cpp:
```diff
@@ -470,14 +474,26 @@ const API::Result<std::string> FileFinderImpl::findRun(const std::string &hintst
+  // Path not found
+  if (!useExtsOnly && uniqueExts.size() == 1) {
+    uniqueExts.pop_back(); // No need to search for missing extension again
+    getUniqueExtensions(exts, uniqueExts);
+    getUniqueExtensions(extensions, uniqueExts);
```
Just in general I think having three vectors, `exts`, `extensions`, `uniqueExts`, is very confusing. If you can think of a way to rename them / restructure things, I think that'd be very helpful.

As far as I can see `extensions` is just assigned the contents of `facility_extensions` so I'm wondering if it even needs to exist? Why not just use `facility_extensions`? (and maybe rename it to `facilityExtensions` while you're at it)

I also think `findRun` is in need of a doc string, `userExtsOnly` is pretty ambiguous. None of this is your fault, just a good time to improve it.

GuiMacielPereira replied:
No, I agree fully, it is incredibly confusing. I'll make it a bit better to understand. You're right, extensions is just the facility extensions, `exts` is just the argument used with `useExtsOnly` and `uniqueExts` are the extensions that are effectively considered. I'll clear it up.

jhaigh0 reviewed (changes requested):
Code is working as expected, good job. I've left some comments about cleaning up `findRun` while we're here making changes.

GuiMacielPereira commented on Framework/API/test/FileFinderTest.h:
```diff
@@ -104,6 +104,8 @@ class FileFinderTest : public CxxTest::TestSuite {
 
     ConfigService::Instance().updateFacilities(m_facFile.path());
     ConfigService::Instance().setString("datacachesearch.directory", "");
+    const std::string propfile = ConfigService::Instance().getDirectoryOfExecutable() + "Mantid.properties";
+    ConfigService::Instance().updateConfig(propfile);
```
I thought this could affect other tests, but they all ran fine.
We just need the user directories, everything else is ignored.
I'll just set the user directories using the api, I agree that is safer and clearer

jhaigh0 replied:
Oh I see, for loading the data. I know there are test data folders which hold data for unit tests specifically, maybe you can find out how those are typically used

GuiMacielPereira replied:
Well, what is happening is that there is a `Mantid.properties` file in the `bin` folder but there is also another properties file in the conda environment folder, which takes priority over the one in the bin folder. When running on Jenkins, the one in the bin folder is used and all is well. When running on my machine, the one in my conda environment is used which causes the missing data directories.

I could set the data directories at the beginning of `FileFinderTest` by hard coding them relative to the bin directory. But if the structure of the build folder changes, then the test would start failing. I also thought about reading the `Mantid.properties` file in the bin folder to look for the data directories, but since `ConfigService` is a singleton, I can't really read the properties file without also updating the properties that the singleton holds, which is what I am already doing anyway with the `updateConfig` call above (unless, of course, I change the ConfigService class a bit to allow for this).

I couldn't find any other test where the data directories are updated in the config service at the beginning of the test, the closest I got was how the config service is updated several times with `updateConfig` in the `ConfigServiceTest`, in the same way that I am doing here.

I think this is the best solution we can do, because it doesn't affect other tests and even when the test is run locally, it doesn't change the properties files after running it (because at the start of workbench the priority defaults to the properties file in the conda environment).

GuiMacielPereira replied:
Or I can just hard code the paths as well, whichever you prefer

GuiMacielPereira committed:
Use structured bindings to unpack pair

Unpacked pair with auto [] and removed unecessary if statements that
checked whether strings are empty.
Adding empty string or not adding it is the same operation.

GuiMacielPereira committed:
Rename variables

Renamed extensions -> facilityExtensions and
uniqueExts -> extensionsToSearch

GuiMacielPereira committed:
Rename arguments of functinos

Renamed exts -> extensionsProvided and
useExtsOnly -> useOnlyExtensionsProvided

GuiMacielPereira committed:
Clean choosing of extensions

Replaced previous standalone if statements with easier to understand
nested if statements.

GuiMacielPereira committed:
Add doc string for findRun

GuiMacielPereira committed:
Add sugestion from code review

GuiMacielPereira committed:
Add comment to clarify purpose

GuiMacielPereira committed:
Fix argument in doc string comment
